src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                            SyntaxFactory.VariableDeclarator(
+                                forEachStatement.Identifier,
+                                argumentList: null,
+                                SyntaxFactory.EqualsValueClause(expression))))),
+                                forEachStatement.Statement).WithAdditionalAnnotations(Formatter.Annotation), queryExpressionProcessingInfo);
+                return new DocumentUpdateInfo(forEachStatement, statements);
+            }
+
+            private DocumentUpdateInfo ConvertIfInToForeachWithoutExtraVariableDeclaration(
+                ForEachStatementSyntax forEachStatement,
+                QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                // before 
+                //  foreach (var a in from a in b where a > 5 select a) 
+                //  { 
+                //      dosomething(a); 
+                //  }
+                // after 
+                //  foreach (var a in b)
+                //  {
+                //      if (a > 5)
+                //      {
+                //          dosomething(a); 
+                //      }
+                //  }
+                var statements = GenerateStatements(
+                    expression => forEachStatement.Statement.WithAdditionalAnnotations(Formatter.Annotation),
+                    queryExpressionProcessingInfo);
+                return new DocumentUpdateInfo(forEachStatement, statements);
+            }
+
+            private bool TryConvertIfInReturnStatement(
+                ReturnStatementSyntax returnStatement,
+                QueryExpressionProcessingInfo queryExpressionProcessingInfo,
+                out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // The conversion requires yield return which cannot be added to lambdas and anonymous method declarations.
+                if (IsWithinImmediateLambdaOrAnonymousMethod(returnStatement))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                var memberDeclarationNode = FindParentMemberDeclarationNode(returnStatement);
+                if (memberDeclarationNode == null)
+                {
+                    throw new ArgumentException("Must bail out on previous steps");
+                }
+
+                var declaredSymbol = _semanticModel.GetDeclaredSymbol(memberDeclarationNode);
+                ITypeSymbol typeSymbol;
+                switch (declaredSymbol.Kind)
+                {
+                    case SymbolKind.Local:
+                        typeSymbol = ((ILocalSymbol)declaredSymbol).Type;
+                        break;
+                    case SymbolKind.Method:
+                        typeSymbol = ((IMethodSymbol)declaredSymbol).ReturnType;
+                        break;
+                    case SymbolKind.Property:
+                        typeSymbol = ((IPropertySymbol)declaredSymbol).Type;
```

Comment:
>typeSymbol = ((IPropertySymbol)declaredSymbol).Type; [](start = 24, length = 52)

This doesn't feel right. We should be getting type from the accessor instead, I think.